lib/linkerd2-trace-context/src/propagation.rs

```diff
@@ -0,0 +1,248 @@
+        }
+        GRPC_TRACE_FIELD_TRACE_ID => {
+            let id = try_split_to(buf, 16)?;
+            trace!(
+                "reading binary trace field {:?}: {:?}",
+                GRPC_TRACE_FIELD_TRACE_ID,
+                id
+            );
+            context.trace_id = id.into();
+        }
+        GRPC_TRACE_FIELD_TRACE_OPTIONS => {
+            let flags = try_split_to(buf, 1)?;
+            trace!(
+                "reading binary trace field {:?}: {:?}",
+                GRPC_TRACE_FIELD_TRACE_OPTIONS,
+                flags
+            );
+            context.flags = flags.into();
+        }
+        id => {
+            return Err(UnknownFieldId(id).into());
+        }
+    };
+    Ok(())
+}
+
+fn increment_grpc_span_id<B>(request: &mut http::Request<B>, context: &mut TraceContext) {
+    let span_id = Id::new(8);
+
+    trace!(message = "incremented span id", %span_id);
+
+    let mut bytes = Vec::<u8>::new();
+
+    // version
+    bytes.push(0);
+
+    // trace id
+    bytes.push(GRPC_TRACE_FIELD_TRACE_ID);
+    bytes.extend(context.trace_id.0.iter());
+
+    // span id
+    bytes.push(GRPC_TRACE_FIELD_SPAN_ID);
+    bytes.extend(span_id.0.iter());
+
+    // trace options
+    bytes.push(GRPC_TRACE_FIELD_TRACE_OPTIONS);
+    bytes.push(context.flags.0);
+
+    let bytes_b64 = base64::encode(&bytes);
+
+    if let Result::Ok(hv) = HeaderValue::from_str(&bytes_b64) {
+        request.headers_mut().insert(GRPC_TRACE_HEADER, hv);
+    } else {
+        warn!("invalid header: {:?}", &bytes_b64);
+    }
+    context.span_id = Some(span_id);
+}
+
+fn unpack_http_trace_context<B>(request: &http::Request<B>) -> Option<TraceContext> {
+    let parent_id = parse_header_id(request, HTTP_SPAN_ID_HEADER)?;
+    let trace_id = parse_header_id(request, HTTP_TRACE_ID_HEADER)?;
+    let flags = match get_header_str(request, HTTP_SAMPLED_HEADER) {
+        Some("1") => Flags(1),
+        _ => Flags(0),
+    };
+    Some(TraceContext {
+        propagation: Propagation::Http,
+        version: Id(vec![0]),
+        trace_id,
+        parent_id,
+        flags,
+        span_id: None,
+    })
+}
+
+fn increment_http_span_id<B>(request: &mut http::Request<B>, context: &mut TraceContext) {
+    let span_id = Id::new(8);
+
+    trace!("incremented span id: {}", span_id);
+
+    let span_str = hex::encode(span_id.as_ref());
+
+    if let Result::Ok(hv) = HeaderValue::from_str(&span_str) {
+        request.headers_mut().insert(HTTP_SPAN_ID_HEADER, hv);
+    } else {
+        warn!("invalid {} header: {:?}", HTTP_SPAN_ID_HEADER, span_str);
+    }
+    context.span_id = Some(span_id);
+}
+
+fn get_header_str<'a, B>(request: &'a http::Request<B>, header: &str) -> Option<&'a str> {
+    let hv = request.headers().get(header)?;
+    hv.to_str()
+        .map_err(|e| warn!("Invalid trace header {}: {}", header, e))
+        .ok()
+}
+
+fn parse_header_id<B>(request: &http::Request<B>, header: &str) -> Option<Id> {
```

Comment:
Maybe something to just keep in mind, but it looks like there is some overlap with this and header retrieval functions in `src/proxy/http/mod.rs`. I don't think we need to consolidate right now, but maybe leave a comment about doing it in the future--maybe when we move some of that into `lib/` as well?